profiler/profiler.go

```diff
@@ -151,25 +161,29 @@ func (p *profiler) enqueueUpload(bat batch) {
 
 // send takes profiles from the output queue and uploads them.
 func (p *profiler) send() {
-	defer close(p.exit)
 	for bat := range p.out {
+		if p.uploadFunc == nil {
+			continue
+		}
 		if err := p.uploadFunc(bat); err != nil {
 			log.Error("Failed to upload profile: %v\n", err)
+
+			if err == errOldAgent {
+				// Turn any future uploads into noops since we know they won't work.
+				p.uploadFunc = nil
```

Comment:
I would prefer we don't use fields which have a purpose as flags to trigger certain behaviour, by setting them to `nil`. This can still race on line 168 without a lock. I don't think this is the way to do this.

I don't have any good ideas. A global atomic variable `connectFailing` which would indicate failure could help you avoid any HTTP calls inside the actual implementation of `p.upload`. Then you could leave this function unchanged. Then it can potentially race in an edge case by doing more than one HTTP request before stopping but that's better than a `nil` panic race. By scoping the changes down to `p.upload` it makes more sense and leaves this function unchanged.

There might be better ideas than what I suggest with the global atomic. Might be ok given we have only one instance. It could also be a field in the profiler.

Reply:
It should also not be the end of the world if the profiler keeps continuing to connect with eventual success? The error logs never get spammy (there's a throttler)